File: v2/9.py




class Solution:
    def isPalindrome(self, x):
        """
        :type x: int
        :rtype: bool
        """
        # 4 star, 不断比较当前数字的首位和末位，如果一样则去除首尾两个数字，继续比较剩下部分的首位和末位
        # 重点在如何得到一个数字的左边和右边的数
        if x < 0:
            return False
        div = 1
        while x / div >= 10:
            div *= 10
        while x > 0:
            left = x // div
            right = x % 10
            if left != right:
                return False
            x %= div
            x //= 10
            div //= 100
        return True

# if __name__ == "__main__":
#     assert Solution().isPalindrome(123) == False
#     assert Solution().isPalindrome(1) == True

File: go/9.go
//package main
//
//import (
//	"fmt"
//	//"strconv"
//)


//将数字转换为字节数组，然后从首尾两边逐一对比，如果首尾不同，则不是回文
//func isPalindrome(x int) bool {
//	str := strconv.Itoa(x)
//	var intArray = []byte(str)
//	left, right := 0, len(intArray) - 1
//	for left <= right {
//		if intArray[left] != intArray[right] {
//			return false
//		} else {
//			left += 1
//			right -= 1
//		}
//	}
//	return true
//}


func isPalindrome(x int) bool {
	if x < 0 {return false}
	div := 1
	for x / div >= 10 {
		div *= 10
	}

	for x > 0 {
		left := x/div
		right := x % 10
		if left != right {
			return false
		}
		x %= div
		x -= right
		x /= 10
		div /= 100
	}
	return true
}



//func main() {
//	rs := isPalindrome(121)
//	fmt.Println(rs)
//}